src/dataframerow/dataframerow.jl

```diff
@@ -337,62 +337,124 @@ end
+        r = row(dfr)
+        for col in _columns(df)
+            # use a barrier function to improve performance
+            pushhelper!(col, r)
+        end
+        for (colname, col) in zip(_names(df), _columns(df))
+            if length(col) != targetrows
+                for col2 in _columns(df)
+                    resize!(col2, nrows)
+                end
+                throw(AssertionError("Error adding value to column :$colname"))
+            end
+        end
+        return df
+    end
+
     if ncols == 0
         for (n, v) in pairs(dfr)
             setproperty!(df, n, fill!(Tables.allocatecolumn(typeof(v), 1), v))
         end
         return df
     end
 
-    current_col = 0
-    try
-        if parent(dfr) === df && index(dfr) isa Index
-            # in this case we are sure that all we do is safe
-            r = row(dfr)
-            for col in _columns(df)
-                # use a barrier function to improve performance
-                pushhelper!(col, r)
+    if cols == :union
+        for (i, colname) in enumerate(_names(df))
+            col = _columns(df)[i]
+            if haskey(dfr, colname)
+                val = dfr[colname]
+            else
+                val = missing
             end
-        else
-            # DataFrameRow can contain duplicate columns and we disallow this
-            # corner case when push!-ing
-            # Only check for equal lengths, as an error will be thrown below if some names don't match
-            if cols === :orderequal
-                if _names(df) != _names(dfr)
-                    msg = "when `cols=:equal` pushed row must have the same column " *
-                          "names and in the same order as the target data frame"
-                    throw(ArgumentError(msg))
-                end
-            elseif cols === :setequal || cols === :equal
-                if cols === :equal
-                    Base.depwarn("`cols=:equal` is deprecated." *
-                                 "Use `:setequal` instead.", :push!)
+            S = typeof(val)
+            T = eltype(col)
+            V = promote_type(S, T)
+            if S <: T || V <: T || !promote
+                push!(col, val)
+            else
+                newcol = Tables.allocatecolumn(V, targetrows)
+                copyto!(newcol, 1, col, 1, nrows)
+                newcol[end] = val
+                _columns(df)[i] = newcol
+            end
+        end
+        for (colname, col) in zip(_names(df), _columns(df))
+            if length(col) != targetrows
+                for col2 in _columns(df)
+                    resize!(col2, nrows)
                 end
-                msg = "Number of columns of `DataFrameRow` does not match that of " *
-                      "target data frame (got $(length(dfr)) and $ncols)."
-                ncols == length(dfr) || throw(ArgumentError(msg))
+                throw(AssertionError("Error adding value to column :$colname"))
```

Comment:
This is already checked in the `catch` block, isn't it?

Reply:
We do not reach `catch` block if `cols == :union`. This error is abnormal, and a most likely cause is that `DataFrame` contained column aliases.